source/brailleDisplayDrivers/albatross/__init__.py

```diff
@@ -0,0 +1,677 @@
+					self._dev.close()
+				self._dev = None
+			if self._kc:
+				self._kc.stop()
+				self._kc = None
+			if self._readQueue:
+				self._readQueue.clear()
+			if self._writeQueue:
+				self._writeQueue.clear()
+			self._readQueue = None
+			self._writeQueue = None
+			self._handleRead = None
+		except Exception:
+			# Terminating anyway
+			pass
+
+	# _initConnection, _initPort, _openPort, _readInitByte and _readSettingsByte
+	# are helper functions to establish connection.
+	# If no connection, Albatross sends continuously INIT_START_BYTE
+	# followed by byte containing various settings like number of cells.
+	def _initConnection(self) -> bool:
+		for i in range(MAX_INIT_SLEEPS):
+			if not self._dev:
+				if not self._initPort(i):
+					continue
+			elif not self._dev.is_open:
+				if not self._openPort(i):
+					continue
+			if self._tryToConnect:
+				self._tryToConnect = False
+			else:
+				log.debug(
+					f"Sleeping {TIMEOUT} seconds before try {i + 1} / {MAX_INIT_SLEEPS}"
+				)
+				time.sleep(TIMEOUT)
+			if not self._readInitByte():
+				if self._invalidSettingsByte:
+					return False
+				else:
+					continue
+			if not self._tryToConnect and self.numCells:
+				return True
+		return False
+
+	# Parameter i is just for loggint retries
+	def _initPort(self, i=MAX_INIT_SLEEPS - 1) -> bool:
+		try:
+			self._dev = serial.Serial(
+				self._currentPort, baudrate=BAUD_RATE, stopbits=serial.STOPBITS_ONE,
+				parity=serial.PARITY_NONE, timeout=TIMEOUT, writeTimeout=WRITE_TIMEOUT)
+			log.debug(f"Port {self._currentPort} initialized")
+			if not self._resetBuffers():
+				if i == MAX_INIT_SLEEPS - 1:
+					return False
+				log(
+					f"sleeping {TIMEOUT} seconds before try {i + 2} / {MAX_INIT_SLEEPS}")
+				time.sleep(TIMEOUT)
+				return False
+			return True
+		except IOError:
+			if i == MAX_INIT_SLEEPS - 1:
+				log.debug(f"Port {self._currentPort} not initialized", exc_info=True)
+				return False
+			log.debug(
+				f"Port {self._currentPort} not initialized, sleeping {TIMEOUT} seconds "
+				f"before try {i + 2} / {MAX_INIT_SLEEPS}", exc_info=True
+			)
+			time.sleep(TIMEOUT)
+			return False
+
+	# Parameter i is just for loggint retries
+	def _openPort(self, i=MAX_INIT_SLEEPS - 1) -> bool:
+		try:
+			self._dev.open()
+			log.debug(f"Port {self._currentPort} opened")
+			if not self._resetBuffers():
+				if i == MAX_INIT_SLEEPS - 1:
+					return False
+				log(
+					f"sleeping {TIMEOUT} seconds before try {i + 2} / {MAX_INIT_SLEEPS}")
+				time.sleep(TIMEOUT)
+				return False
+			return True
+		except IOError:
+			if i == MAX_INIT_SLEEPS - 1:
+				log.debug(f"Port {self._currentPort} not opened", exc_info=True)
+				return False
+			log.debug(
+				f"Port {self._currentPort} not opened, sleeping {TIMEOUT} seconds "
+				f"before try {i + 2} / {MAX_INIT_SLEEPS}", exc_info=True
+			)
+			time.sleep(TIMEOUT)
+			return False
+
+	def _readInitByte(self) -> bool:
+		# Strange but very rarely in_waiting causes exception.

```

Comment:
What causes this?

Reply:
> What causes this?

I am not sure. Maybe communication failure?